thavalon-server/src/lobby/mod.rs

```diff
@@ -0,0 +1,270 @@
+        tokio::task::spawn(async move {
+            while let Some(msg) = from_player_client.next().await {
+                let msg = match msg {
+                    Ok(message) => message.to_str().unwrap().to_owned(),
+                    Err(e) => {
+                        // It's not entirely clear what an error here means, so
+                        // probably best to just close the entire connection.
+                        log::error!(
+                            "An error occurred while waiting for messages from player {}. {}",
+                            client_id,
+                            e
+                        );
+                        break;
+                    }
+                };
+
+                log::debug!("Received message {} from player {}.", msg, client_id);
+                let action: Action = match serde_json::from_str(&msg) {
+                    Ok(action) => action,
+                    Err(e) => {
+                        log::warn!(
+                            "Could not deserialize message from player {}. Message: {}",
+                            client_id,
+                            msg
+                        );
+                        // TODO: For now, just kill the connection.
+                        break;
+                    }
+                };
+                to_game.send(action).await.unwrap();
+            }
+        });
+
+        // Spawn a new task to forward messages from the game to the player.
+        // The lock may appear weird, but it is needed to get send + sync on
+        // the channel. The channel will only be held by one person ever.
+        tokio::task::spawn(async move {
+            while let Some(msg) = from_game.lock().await.recv().await {
+                let msg = serde_json::to_string(&msg).expect("Could not serialize game message.");
+                let msg = ws::Message::text(msg);
+                if let Err(e) = to_player_client.send(msg).await {
+                    log::warn!("Connection lost. {}", e);
+                    break;
+                }
+            }
+        });
+
+        LobbyResponse::Standard(Ok(()))
+    }
+
+    /// Begins a loop for the lobby to listen for incoming commands.
+    /// This function should only return when the game ends or when a fatal
+    /// error occurs.
+    async fn listen(mut self) {
+        while let Some(msg) = self.receiver.recv().await {
+            let (msg_contents, result_channel) = msg;
+            let results = match msg_contents {
+                LobbyCommand::AddPlayer {
+                    player_id,
+                    display_name,
+                } => self.add_player(player_id, display_name).await,
+
+                LobbyCommand::GetFriendCode => self.get_friend_code(),
+                LobbyCommand::IsPlayerRegistered { player_id } => {
+                    LobbyResponse::IsPlayerRegistered(self.players.contains_key(&player_id))
+                }
```

Comment:
What do you think of collapsing `IsPlayerRegistered` into `ConnectClientChannels`? `ConnectClientChannels` could return an error if the player isn't in the game, and then we'd need fewer messages

Reply:
I actually made all of these changes and then decided to change it back. The reason being that without checking if the player is registered, we fully upgrade the websocket connection any time the game alone exists. I'd imagine this involves a round trip to the client and more of a performance hit, especially considering we're creating new threads on the server. We'd then immediately close this connection and all of the threads upon checking and seeing the player isn't in the game, instead of just rejecting the connection outright.